prov/gni/src/gnix_av.c

```diff
@@ -36,13 +36,449 @@
+		break;
+	default:
+		ret = -FI_EINVAL;
+		break;
+	}
+
+err:
+	return ret;
+}
+
+/*
+ * TODO: Fix implementation for FI_AV_MAP so it's actually stored and looked up
+ * rather than recreated each time.
+ */
+static int gnix_av_insert(struct fid_av *av, const void *addr, size_t count,
+			  fi_addr_t *fi_addr, uint64_t flags, void *context)
+{
+	struct gnix_fid_av *int_av;
+	int ret = FI_SUCCESS;
+
+	if (!av || !addr) {
+		ret = -FI_EINVAL;
+		goto err;
+	}
+
+	/*
+	 * fi_addr parameter may be NULL only if the count is 0. Otherwise it
+	 * must reference an array of fi_addr_t.
+	 */
+	if (!fi_addr && count) {
+		ret = -FI_EINVAL;
+		goto err;
+	}
+
+	int_av = container_of(av, struct gnix_fid_av, av_fid);
+
+	switch (int_av->type) {
+	case FI_AV_TABLE:
+		ret =
+		    table_insert(int_av, addr, count, fi_addr, flags, context);
+		break;
+	case FI_AV_MAP:
+		ret = map_insert(int_av, addr, count, fi_addr, flags, context);
+		break;
+	default:
+		ret = -FI_EINVAL;
+		break;
+	}
+
+err:
+	return ret;
+}
+
+static int gnix_av_remove(struct fid_av *av, fi_addr_t *fi_addr, size_t count,
+			  uint64_t flags)
+{
+	struct gnix_fid_av *int_av;
+	int ret = FI_SUCCESS;
+
+	if (!av || !fi_addr) {
+		ret = -FI_EINVAL;
+		goto err;
+	}
+
+	int_av = container_of(av, struct gnix_fid_av, av_fid);
+
+	switch (int_av->type) {
+	case FI_AV_TABLE:
+		ret = table_remove(int_av, fi_addr, count, flags);
+		break;
+	case FI_AV_MAP:
+		ret = map_remove(int_av, fi_addr, count, flags);
+		break;
+	default:
+		ret = -FI_EINVAL;
+		break;
+	}
+
+err:
+	return ret;
+}
+
+/*
+ * Given an address pointed to by addr, stuff a string into buf representing:
+ * device_addr:cdm_id
+ * where device_addr and cdm_id are represented in hexadecimal.
+ */
+static const char *gnix_av_straddr(struct fid_av *av, const void *addr,
+				   char *buf, size_t *len)
+{
+	char int_buf[64];
+	int size;
+
+	const struct gnix_address *gnix_addr = addr;
+
+	size =
+	    snprintf(int_buf, sizeof(int_buf), "0x%08" PRIx32 ":0x%08" PRIx32,
+		     gnix_addr->device_addr, gnix_addr->cdm_id);
+
+	snprintf(buf, *len, "%s", int_buf);
+	*len = size + 1;
+
+	return buf;
+}
+
+/*
+ * TODO: Free memory for data structures when FI_AV_MAP is fully supported.
+ */
+static int gnix_av_close(fid_t fid)
+{
+	struct gnix_fid_av *av;
+	av = container_of(fid, struct gnix_fid_av, av_fid.fid);
+	if (av->table) {
+		free(av->table);
+	}
+	free(av);
+	return 0;
+}
+
+/*
+ * TODO: Support shared named AVs.
+ */
 int gnix_av_open(struct fid_domain *domain, struct fi_av_attr *attr,
 		 struct fid_av **av, void *context)
 {
-	/* TODO: implement this puppy */
-	return -FI_ENOSYS;
+	struct gnix_fid_domain *int_dom;
+	struct gnix_fid_av *int_av;
+
+	enum fi_av_type type = FI_AV_TABLE;
+	size_t count = 128;
+
+	int_dom = container_of(domain, struct gnix_fid_domain, domain_fid);
+
+	int_av = calloc(1, sizeof(*int_av));
+	if (!int_av) {
+		return -FI_ENOMEM;
+	}
+
+	if (attr) {
+		if (gnix_verify_av_attr(attr)) {
```

Comment:
I meant in this particular error case.  Seems like it would get lost here, but maybe I'm missing something.